scenario/scripts/snapshot.py

```diff
@@ -0,0 +1,752 @@
+import json
+import logging
+import os
+import re
+import tempfile
+from dataclasses import asdict
+from enum import Enum
+from itertools import chain
+from pathlib import Path
+from subprocess import CalledProcessError, check_output, run
+from textwrap import dedent
+from typing import Any, BinaryIO, Dict, Iterable, List, Optional, TextIO, Tuple, Union
+
+import ops.pebble
+import typer
+import yaml
+
+from scenario.state import (
+    Address,
+    BindAddress,
+    Container,
+    Model,
+    Mount,
+    Network,
+    Relation,
+    State,
+    Status,
+)
+
+logger = logging.getLogger("snapshot")
+
+JUJU_RELATION_KEYS = frozenset({"egress-subnets", "ingress-address", "private-address"})
+JUJU_CONFIG_KEYS = frozenset({})
+SNAPSHOT_TEMPDIR_ROOT = (Path(os.getcwd()).parent / "snapshot_storage").absolute()
+
+
+class SnapshotError(RuntimeError):
+    pass
+
+
+class InvalidTarget(SnapshotError):
+    pass
+
+
+class InvalidModel(SnapshotError):
+    pass
+
+
+class Target(str):
+    def __init__(self, unit_name: str):
+        super().__init__()
+        app_name, _, unit_id = unit_name.rpartition("/")
+        if not app_name or not unit_id:
+            raise InvalidTarget(f"invalid unit name: {unit_name!r}")
+        self.unit_name = unit_name
+        self.app_name = app_name
+        self.unit_id = int(unit_id)
+        self.normalized = f"{app_name}-{unit_id}"
+
+
+def _try_format(string):
+    try:
+        import black
+
+        try:
+            return black.format_str(string, mode=black.Mode())
+        except black.parsing.InvalidInput as e:
+            logger.error(f"error parsing {string}: {e}")
+            return string
+    except ModuleNotFoundError:
+        logger.warning("install black for formatting")
+        return string
```

Comment:
it's not about code injection, it really is about pretty-printing (formatting) it. 
Issue is: str(State) will return valid python code, but it looks ugly because there's no newlines, no indents, etc... it's just a single very long line.

So you have a choice: output as json (which we can automatically format and indent), output as python (valid, but ugly), or install black and output as python AND pretty.

Reply:
Hmm... I guess what I am not clear about is why you want to force people to install black manually instead of just having it available by default with ops-scenario? How many people do you think will notice the warning message in the logger, especially if the have their logger set to only report errors and not debug information? I feel like it would just be easier to just have black be installed alongside scenario when I `pip install ...` rather than hope people seeing the warning message in their logs.